### PR TITLE
Initial DAO configuration part 2

### PR DESCRIPTION
**Pull request by whoabuddy**
This PR follows up on #1 with some additional tests and once merged will represent the initial implementation.

### CODE REVIEW

File: contracts/extensions/ccd002-treasury-mia.clar [hunk 1]
```diff
@@ -62,71 +70,106 @@
...
+      recipient: TREASURY
+    })
     (ok true)
   )
 )
 
-(define-public (transfer (amount uint) (recipient principal))
+;; Withdraw functions
+
+(define-public (withdraw-stx (amount uint) (recipient principal))
```
Comment (ryanwaits) [inline contracts/extensions/ccd002-treasury-mia.clar:123]: nice name change here, so as not to confuse with stx-transfer?

**Comment by radicleart** [comment]
General comments and questions based on a manual comparison of the base dao, vs xDAO contracts.

1. General:  The differences are all formatting related except for a public function `direct-execute` in ccd01. Maybe worth considering keeping the external interface consistent for xDAO offspring projects as this may lead to easier reuse of of xDAO tooling / api layers?

2. ccd002 - only difference between mia and nyc reasury contracts is line 1. Maybe less maintenance to hold the name in a const and deploy the same contract twice?

3. ccd002 - remove reference to SIP013 in synopsis section.

4. ccd002 - will sip013 be supported if the SIP is ratified ?

5. ccd002 - change whitelist to allowlist to keep the language inclusive?

6. ccd002 - is whitelisting on transfers necessary - given the transfer can only be called by the DAO?

7. ccd002 - the deposit functionality - is the motivation to prevent unwanted assets being transferred to the DAO ?

**Comment by whoabuddy** [comment]
> General comments and questions based on a manual comparison of the base dao, vs xDAO contracts.

This is great, thank you for making the time!

> General: The differences are all formatting related except for a public function direct-execute in ccd01. Maybe worth considering keeping the external interface consistent for xDAO offspring projects as this may lead to easier reuse of of xDAO tooling / api layers?

There were definitely some formatting changes that I'm in favor of, inspired by @LNow's [work here](https://github.com/LNow/clarity-notes-archive/blob/main/ideas/coding-standards.md) and what we already have with CityCoins contracts.

Regarding the external interface, that's a good point, there were actually a few other small changes based on the previous auth contracts defined in [CCIP-007](https://github.com/citycoins/governance/blob/main/ccips/ccip-007/ccip-007-citycoins-auth.md) and [CCIP-010](https://github.com/citycoins/governance/blob/main/ccips/ccip-010/ccip-010-citycoins-auth-v2.md):

- sunset time increased to ~6 months
- `executive` or `executive-team` replaced with `approver`
- `executive-action` replaced with `direct-execute`
- maps, vars, constants, functions and errors are edited a bit as a result

How do you imagine the tooling or API layers would interact with CityCoins? My vote would be to keep this as-is and focus on that type of integration for future proposal mechanisms, e.g. if we [recreate CCIP-011 as an extension](https://github.com/citycoins/governance/blob/main/ccips/ccip-011/ccip-011-citycoins-stacked-tokens-voting.md).

> ccd002 - only difference between mia and nyc reasury contracts is line 1. Maybe less maintenance to hold the name in a const and deploy the same contract twice?

In this case I was trying to follow the same naming convention as other contracts, with stx.eco you were able to pull this information and display it, so the (loose) standard would be for each contract to have:

- title
- version
- synopsis
- description

Curious how you think holding the name in a const would make maintenance easier, can you expand?

> ccd002 - remove reference to SIP013 in synopsis section.

I believe [this was done](https://github.com/citycoins/protocol/pull/2/files#diff-b6f3577a731171c00b1e65925de4f4ef847a1a0db5a6fceb763bba55ddcf21c3) but def will make sure!

> ccd002 - will sip013 be supported if the SIP is ratified ?

I don't see why not, there should be an upgrade procedure for treasuries that support new assets and future SIPs - I've had some thoughts about using `at-block` and chained queries to see historical balance at past block heights.

Shared some thoughts here: https://github.com/citycoins/protocol/issues/6

> ccd002 - change whitelist to allowlist to keep the language inclusive?

That's fine, I'm not picky about that one! Fixed in 56e7bf2.

> ccd002 - is whitelisting on transfers necessary - given the transfer can only be called by the DAO?

I believe this idea originally came from @LNow's Clarity security notes around [securing as-contract](https://github.com/LNow/clarity-notes-archive/blob/main/security/as-contract.md) and [functions that use traits](https://github.com/LNow/clarity-notes-archive/blob/main/security/traits.md).

The extra step makes the DAO explicitly define what token contracts it supports, giving an extra opportunity to inspect the code and make sure the trait is applied correctly.

> ccd002 - the deposit functionality - is the motivation to prevent unwanted assets being transferred to the DAO?

No, more to provide a consistent interface that matches the withdraw functions for the supported assets, but I have debated where this would be used.

I could see it used with other extensions or parts of the protocol that interact with the treasuries, where calling this function could be better than directly inputting the principal.

**Comment by radicleart** [comment]
`There were definitely some formatting changes that I'm in favor of, inspired by @LNow's [work here](https://github.com/LNow/clarity-notes-archive/blob/main/ideas/coding-standards.md) and what we already have with CityCoins contracts.`
👍 (yep much nicer esp. when copying things like the error constants into js test files!)

`How do you imagine the tooling or API layers would interact with CityCoins? My vote would be to keep this as-is and focus on that type of integration for future proposal mechanisms, e.g. if we [recreate CCIP-011 as an extension](https://github.com/citycoins/governance/blob/main/ccips/ccip-011/ccip-011-citycoins-stacked-tokens-voting.md).`

Not so important for this function but generally sticking with a convention for naming public methods (for projects building on xDAO) will make it easier to abstract common indexing layer for fast access to queries like voting history per proposal, active extensions, views of proposals I voted for and so on. I can see this might be tricky given the number of ways the DAO can grow - similar to what Hiro are trying to provide for indexing NFT data.

`Curious how you think holding the name in a const would make maintenance easier, can you expand?
`

Just that if you had a thousand cities you'd and up with a thousand functionally equivalent contracts in the repo this way.

` The extra step makes the DAO explicitly define what token contracts it supports, giving an extra opportunity to inspect the code and make sure the trait is applied correctly.
`

👍 yes this sounds familiar - thanks for the explanation.


`No, more to provide a consistent interface that matches the withdraw functions for the supported assets, but I have debated where this would be used.
`
👍

**Comment by whoabuddy** [comment]
> Not so important for this function but generally sticking with a convention for naming public methods (for projects building on xDAO) will make it easier to abstract common indexing layer for fast access to queries like voting history per proposal, active extensions, views of proposals I voted for and so on. I can see this might be tricky given the number of ways the DAO can grow - but the idea is to be able to build something similar to what Hiro are trying to provide for indexing NFT data.

This makes sense, one idea we had was to add some endpoints to the [CC API](https://api.citycoins.co/docs) for proposal queries, much like the contract bits we have in place now. It'd also be interesting to look at how much other projects already using xDAO have evolved (StackerDAOs, ALEX, others?).

> Just that if you had a thousand cities you'd and up with a thousand functionally equivalent contracts in the repo this way.

There are definitely some pros/cons here, I like the approach of having separate treasuries per city, and even further, treating treasuries like an account where each city could have one for: mining proceeds, stacking proceeds, spending accounts, etc.

That way proposals would have to clearly define the control of funds, and eventually cities could deploy an extension with control over a subset of contracts, such as the treasuries. The big picture is there, but we're committed to starting small and simple :)

It does open up some questions about managing a larger amount of contracts and subsequent versioning, we saw some of those issues first-hand in the original CityCoins implementations, and hope to make it much easier to have  the best of both worlds in this iteration.

**Comment by radicleart** [comment]
`It does open up some questions about managing a larger amount of contracts and subsequent versioning, we saw some of those issues first-hand in the original CityCoins implementations, and hope to make it much easier to have the best of both worlds in this iteration.
`

Just to clarify - I meant one contract in github but still deploy a separate contract per treasury/city - i.e. set the contract name at deploy time rather than in github. It might be better to keep the contract body unchanged on deployment, so it doesn't mention the city at all in comments but uses the contract name to identify the 3 letter city code.

**Comment by whoabuddy** [comment]
> Just to clarify - I meant one contract in github but still deploy a separate contract per treasury/city - i.e. set the contract name at deploy time rather than in github. It might be better to keep the contract body unchanged on deployment, so it doesn't mention the city at all in comments or as a const but uses the contract name to identify the 3 letter city code.

Got it, this makes sense, thanks!

**Comment by whoabuddy** [comment]
Finished simplifying the treasury contract into a single extension, where the first two on-chain versions can use the city symbol in their contract name to differentiate what they're used for.

This should be ready to merge! Codecov is being funky, see #7.

**Review by ryanwaits — APPROVED** [review]
nice refactor here , much cleaner

File: tests/extensions/ccd001-direct-execute.test.ts [hunk 2]
```diff
@@ -1,15 +1,14 @@
 import { Account, assertEquals, Clarinet, Chain } from "../../utils/deps.ts";
-import { BASE_DAO, EXTENSIONS, PROPOSALS } from "../../utils/common.ts";
+import { EXTENSIONS, PROPOSALS } from "../../utils/common.ts";
 import { CCD001DirectExecute } from "../../models/extensions/ccd001-direct-execute.model.ts";
 
-const ccd001DirectExecute = new CCD001DirectExecute();
-
```
Comment (moodmosaic) [inline tests/extensions/ccd001-direct-execute.test.ts:6]: 👍

@whoabuddy, with the forthcoming fuzz/prop stuff the following will be also possible:

image

**Pull request merged**